design/bmc-address.md

```diff
@@ -0,0 +1,126 @@
+<!--
+ This work is licensed under a Creative Commons Attribution 3.0
+ Unported License.
+
+ http://creativecommons.org/licenses/by/3.0/legalcode
+-->
+
+# bmc-address
+
+## Status
+
+implementable
+
+## Table of Contents
+
+<!--ts-->
+   * [bmc-address](#bmc-address)
+      * [Status](#status)
+      * [Table of Contents](#table-of-contents)
+      * [Summary](#summary)
+      * [Motivation](#motivation)
+         * [Goals](#goals)
+         * [Non-Goals](#non-goals)
+      * [Proposal](#proposal)
+         * [Implementation Details/Notes/Constraints](#implementation-detailsnotesconstraints)
+         * [Risks and Mitigations](#risks-and-mitigations)
+      * [Design Details](#design-details)
+         * [Work Items](#work-items)
+         * [Dependencies](#dependencies)
+         * [Test Plan](#test-plan)
+         * [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy)
+         * [Version Skew Strategy](#version-skew-strategy)
+      * [Alternatives](#alternatives)
+      * [References](#references)
+
+<!-- Added by: dhellmann, at: 2019-03-04T09:33-0500 -->
+
+<!--te-->
+
+## Summary
+
+This document explains the way users will provide the network location
+of the management controller on a host.
+
+## Motivation
+
+We need to document how we are going to specify the location of a BMC,
+including how to tell its type, IP address, port number, and
+potentially the path.
+
+### Goals
+
+1. To agree on an address specification system, including assumptions
+   that can be made when parts of an address are left out.
+
+### Non-Goals
+
+1. To list the scheme to use for every type of controller.
+2. To specify the user interface for entering address information.
+
+## Proposal
+
+### Implementation Details/Notes/Constraints
+
+For each BMC, we need to know the type, IP, port, and optionally the
+path to use to communicate with the controller.
+
+We can collect all of this information using a single "address" field,
+if we allow that field to contain partial or full URLs.
+
+For each type of controller, we can assume we know the URL scheme used
+to communicate with it (HTTP, HTTPS, etc.). Therefore the scheme in a
+URL provided by the user is redundant, and we can use that portion of
+the URL to specify the controller type. For example:
+
+    ipmi://192.168.111.1
+
+Initially, we will only support IPMI controllers, so we do not need
+users to specify the type.  If the field only contains a network
+address, we can assume that the controller uses IPMI on the standard
+port, 623. Therefore this would be equivalent to the previous example:
+
+    192.168.111.1
+
+### Risks and Mitigations
+
+One risk in this approach is that we would need to tell users how to
```

Comment:
Is the concern for risk when we get to the UI piece? We could have a smart text input that prepends the box with "ipmi://" and appends ":623" at the end allowing the user to simply provide the BMC IP. Eventually, it might make sense to all the user to build that out more freely as we support more types and different ports?

Reply:
I was concerned with the user thinking "URL" and putting "https://" when they should put IPMI, so having the text field stick in the extra parts might make sense. It won't need to do that for the backend, though (we have to protect against bad data coming in through the API anyway), so only implement it if you think it helps with usability in the UI.